File: airmon.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <vector>
#include <cstdlib>
#include <Windows.h>
#include <Urlmon.h>

#pragma comment(lib, "urlmon.lib")

// Function to check if monitor mode is supported on any adapters
std::vector<std::string> checkMonitorModeSupport() {
    std::vector<std::string> adapters;
    std::string line;
    std::ifstream driverOutput("drivers.txt");

    if (driverOutput.is_open()) {
        while (getline(driverOutput, line)) {
            if (line.find("Name") != std::string::npos) {
                adapters.push_back(line.substr(line.find(":") + 1));
            }
        }
        driverOutput.close();
    }

    return adapters;
}

// Function to download and install Npcap
void downloadNpcap() {
    const std::string url = "https://nmap.org/npcap/dist/npcap-1.79.exe";
    const std::string filename = "npcap.exe";

    std::cout << "[+] Downloading Npcap from " << url << std::endl;
    HRESULT result = URLDownloadToFileA(NULL, url.c_str(), filename.c_str(), 0, NULL);
    if (result == S_OK) {
        std::cout << "[+] Download complete. Installing Npcap..." << std::endl;
        system("npcap.exe /S"); // Silent install
        std::cout << "[+] Npcap installed successfully!" << std::endl;
    } else {
        std::cout << "[-] Error downloading Npcap." << std::endl;
    }
}

// Function to enable monitor mode for a specific adapter
void enableMonitorMode(const std::string& adapterName) {
    std::cout << "[+] Enabling monitor mode for adapter: " << adapterName << std::endl;
    std::string command = "wlanhelper.exe " + adapterName + " mode monitor";
    system(command.c_str());
    std::cout << "[+] Monitor Mode enabled!" << std::endl;
}

// Function to disable monitor mode for a specific adapter
void disableMonitorMode(const std::string& adapterName) {
    std::cout << "[+] Disabling monitor mode for adapter: " << adapterName << std::endl;
    std::string command = "wlanhelper.exe " + adapterName + " mode managed";
    system(command.c_str());
    std::cout << "[+] Monitor Mode disabled!" << std::endl;
}

// Function to list adapters in monitor mode
std::vector<std::string> listMonitorModeAdapters() {
    std::vector<std::string> monitorAdapters;
    std::string line;
    std::ifstream monitorOutput("monitor_mode.txt");

    if (monitorOutput.is_open()) {
        while (getline(monitorOutput, line)) {
            if (line.find("Mode") != std::string::npos && line.find("Monitor") != std::string::npos) {
                std::string adapterName = line.substr(line.find(":") + 1);
                monitorAdapters.push_back(adapterName);
            }
        }
        monitorOutput.close();
    }

    return monitorAdapters;
}

// Function to prompt user for an adapter choice
int getAdapterChoice(const std::vector<std::string>& adapters) {
    std::cout << "[+] Available adapters:" << std::endl;
    for (int i = 0; i < adapters.size(); ++i) {
        std::cout << i + 1 << ". " << adapters[i] << std::endl;
    }

    int choice;
    std::cout << "[+] Choose an adapter by number: ";
    std::cin >> choice;

    if (choice >= 1 && choice <= adapters.size()) {
        return choice - 1;
    } else {
        std::cout << "[-] Invalid choice." << std::endl;
        return -1;
    }
}

int main(int argc, char* argv[]) {
    if (argc < 2) {
        std::cout << "Usage: %s start|kill" << std::endl;
        return 1;
    }

    std::string command = argv[1];

    if (command == "start") {
        std::cout << "[+] Checking compatible adapters for monitor mode..." << std::endl;
        std::vector<std::string> adapters = checkMonitorModeSupport();

        if (adapters.empty()) {
            std::cout << "[-] No compatible adapters found that support monitor mode." << std::endl;
            return 1;
        }

        int adapterChoice = getAdapterChoice(adapters);
        if (adapterChoice == -1) return 1;

        std::string adapterName = adapters[adapterChoice];
        std::cout << "[+] Selected adapter: " << adapterName << std::endl;

        char installNpcap;
        std::cout << "[+] Do you want to install Npcap? (y/n): ";
        std::cin >> installNpcap;

        if (installNpcap == 'y' || installNpcap == 'Y') {
            downloadNpcap();
        }

        enableMonitorMode(adapterName);

    } else if (command == "kill") {
        std::cout << "[+] Listing adapters in monitor mode..." << std::endl;
        std::vector<std::string> monitorAdapters = listMonitorModeAdapters();

        if (monitorAdapters.empty()) {
            std::cout << "[-] No adapters are in monitor mode." << std::endl;
            return 1;
        }

        int adapterChoice = getAdapterChoice(monitorAdapters);
        if (adapterChoice == -1) return 1;

        std::string adapterName = monitorAdapters[adapterChoice];
        std::cout << "[+] Selected adapter: " << adapterName << std::endl;

        disableMonitorMode(adapterName);

    } else {
        std::cout << "Usage: %s start|kill" << std::endl;
        return 1;
    }

    return 0;
}

File: airmon.py
import argparse
import os
import subprocess
import urllib.request
import zipfile
import tempfile

NPCAP_URL = "https://nmap.org/npcap/dist/npcap-1.70.exe"

def download_and_install_npcap():
    # Download npcap installer
    print("Downloading npcap...")
    installer_path = os.path.join(tempfile.gettempdir(), "npcap-1.70.exe")
    urllib.request.urlretrieve(NPCAP_URL, installer_path)
    print("Installing npcap...")
    # Install npcap silently
    subprocess.run([installer_path, '/S'], check=True)
    # Add npcap to system PATH
    npcap_path = r"C:\Program Files\Npcap"
    if npcap_path not in os.environ["PATH"]:
        os.environ["PATH"] += os.pathsep + npcap_path
        print("Npcap installed and added to PATH.")
    else:
        print("Npcap is already in PATH.")

def list_adapters():
    # This function lists all network adapters that support monitor mode
    print("Listing compatible adapters that support monitor mode:")
    result = subprocess.run(['npcap', '-WiFi'], capture_output=True, text=True)
    print(result.stdout)

def enable_monitor_mode(adapter_name):
    # This function enables monitor mode on the selected adapter
    print(f"Enabling monitor mode on adapter: {adapter_name}")
    result = subprocess.run(['npcap', '-i', adapter_name, '--set', 'monitor-mode', '1'], capture_output=True, text=True)
    if result.returncode == 0:
        print("Monitor Mode enabled!")
    else:
        print(f"Failed to enable monitor mode: {result.stderr}")

def disable_monitor_mode(adapter_name):
    # This function disables monitor mode on the selected adapter
    print(f"Disabling monitor mode on adapter: {adapter_name}")
    result = subprocess.run(['npcap', '-i', adapter_name, '--set', 'monitor-mode', '0'], capture_output=True, text=True)
    if result.returncode == 0:
        print("Monitor Mode disabled!")
    else:
        print(f"Failed to disable monitor mode: {result.stderr}")

def main():
    parser = argparse.ArgumentParser(description="Manage monitor mode on WiFi adapters using npcap.")
    subparsers = parser.add_subparsers(dest='command', required=True)

    start_parser = subparsers.add_parser('start', help='Enable monitor mode on a selected adapter')
    start_parser.add_argument('adapter', type=str, help='The name of the adapter to set to monitor mode')

    kill_parser = subparsers.add_parser('kill', help='Disable monitor mode on a selected adapter')
    kill_parser.add_argument('adapter', type=str, help='The name of the adapter to disable monitor mode')

    args = parser.parse_args()

    # Check if npcap is installed, if not download and install it
    try:
        subprocess.run(['npcap', '-v'], check=True, capture_output=True)
    except FileNotFoundError:
        download_and_install_npcap()

    if args.command == 'start':
        list_adapters()
        enable_monitor_mode(args.adapter)
    elif args.command == 'kill':
        disable_monitor_mode(args.adapter)

if __name__ == '__main__':
    main()
